Create new Dir and Run Trimgalore using Dac* fastq files

```console
$ mkdir 112019
$ cd 112019
$ mv DacBet_1.fastq DacBet_2.fastq 112019/
$ trim_galore -j 4 -q 30 --paired DacBet_1.fastq DacBet_2.fastq
```
*These raw and galored files will be used by IDBA to assembly*

Create an Idba directory and subdirectories

```console
$ mkdir Idba
$ cd Idba/
$ mkdir Raw_reads
$ cd Raw_reads/
```
Copy raw reads

```console
$ cp ../../DacBet_1.fastq .
$ cp ../../DacBet_2.fastq .
```

Transform fastq to fasta and merge 

```console
$ ~/bin/idba/bin/fq2fa --merge DacBet_1.fastq DacBet_2.fastq DacFasta.fasta
```
Assembly using idba_ud with 4 cpus

```console
$ ~/bin/idba/bin/idba_ud -r DacFasta.fasta -o DacRawIdba --num_threads 4
```
Perform basic stats

```console
$ cd DacRawIdba/
$ perl ../../../Genome_Assembly_lecture/Scripts/assembly.stats.pl contig.fa
Sample_ID	Genome	Contigs	Mean	Median	N50	Largest	GC(%)	N_count	N(%)	Gap_count
contig.fa	3548926	150	23659	15554	37056	153869	62.73		0.00	0
```

Using other script

```console
$ ../../../Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats contig.fa
stats for contig.fa
sum = 3548926, n = 150, ave = 23659.51, largest = 153869
N50 = 37056, n = 28
N60 = 30878, n = 38
N70 = 25661, n = 51
N80 = 19924, n = 66
N90 = 12819, n = 88
N100 = 348, n = 150
N_count = 0
Gaps = 0
```

Now the stats using scaffolds

```console
$ ../../../Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats scaffold.fa 
stats for scaffold.fa
sum = 3551975, n = 167, ave = 21269.31, largest = 153869
N50 = 42221, n = 25
N60 = 34843, n = 34
N70 = 29493, n = 45
N80 = 22033, n = 59
N90 = 13599, n = 79
N100 = 100, n = 167
N_count = 147
Gaps = 4
```

What happen if we use the TrimGalored reads?

```console
$ ~/bin/idba/bin/fq2fa --merge DacBet_1_val_1.fq DacBet_2_val_2.fq DacGalore.fasta
$ nohup ~/bin/idba/bin/idba_ud -r DacGalore.fasta -o DacGaloreIdba --num_threads 4 &
$ cd DacGaloreIdba/
$ ../../../Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats contig.fa
stats for contig.fa
sum = 3548654, n = 150, ave = 23657.69, largest = 153869
N50 = 37057, n = 28
N60 = 30878, n = 38
N70 = 25661, n = 51
N80 = 19924, n = 66
N90 = 12819, n = 88
N100 = 348, n = 150
N_count = 0
Gaps = 0
```
*Remember nohup and & for working command in the background* 

No we trimm those small contings (< 1000 nt) using the perl script trimm_len.pl

```console
$ ../../../Genome_Assembly_lecture/Scripts/trimm_len.pl contig.fa 1000 > contig.1000.fa
```

Let's look the stats

```console
$ ../../../Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats contig.1000.fa
stats for contig.1000.fa
sum = 3544292, n = 142, ave = 24959.80, largest = 153869
N50 = 37415, n = 27
N60 = 30878, n = 38
N70 = 25699, n = 50
N80 = 19924, n = 66
N90 = 12819, n = 88
N100 = 1144, n = 142
N_count = 0
Gaps = 0
```

Now let's modify the k-mer steep iteration using 10 and 30

```console
$ nohup ~/bin/idba/bin/idba_ud --step 10 -r DacGalore.fasta -o DacGaloreIdba.kmerSteep10 --num_threads 4 &
$ nohup ~/bin/idba/bin/idba_ud --step 30 -r DacGalore.fasta -o DacGaloreIdba.kmerSteep30 --num_threads 4 &
```

And check the stats

```console
$ cd DacGaloreIdba.kmerSteep30
$ ../../../Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats contig.fa
stats for contig.fa
sum = 3548340, n = 151, ave = 23498.94, largest = 153869
N50 = 37057, n = 28
N60 = 30421, n = 39
N70 = 25661, n = 51
N80 = 19924, n = 67
N90 = 12819, n = 89
N100 = 348, n = 151
N_count = 0
Gaps = 0
$ cd ..
$ cd DacGaloreIdba.kmerSteep10/
$ ../../../Genome_Assembly_lecture/Scripts/assembly-stats/assembly-stats contig.fa
stats for contig.fa
sum = 3549827, n = 149, ave = 23824.34, largest = 153869
N50 = 37488, n = 27
N60 = 31088, n = 37
N70 = 25699, n = 50
N80 = 19950, n = 65
N90 = 12126, n = 88
N100 = 355, n = 149
N_count = 0
Gaps = 0
```
